clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,41 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty)) {
+    return (CATy->getSize() == 0);
+  }
+  return false;
+}
+
+void Sema::checkSYCLVarDeclIfInKernel(VarDecl *Var) {
+  // not all variable types supported in kernel contexts
+  // if not we record a deferred diagnostic.
+  assert(getLangOpts().SYCLIsDevice &&
+         "Should only be called during SYCL compilation");
+  QualType Ty = Var->getType();
+  SourceRange Loc = Var->getLocation();
+
+  // __int128, __int128_t, __uint128_t
+  if (Ty->isSpecificBuiltinType(BuiltinType::Int128) ||
+      Ty->isSpecificBuiltinType(BuiltinType::UInt128))
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+        << Ty.getUnqualifiedType().getCanonicalType().getAsString();
+
+  // QuadType __float128
+  if (Ty->isSpecificBuiltinType(BuiltinType::Float128) &&
+      !Context.getTargetInfo().hasFloat128Type())
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+        << "__float128";
```

Comment:
Do BuiltinType::Float128 here.

Reply:
`BuiltinType::Float128` is just an enumeration value.  But I could use `Ty.getUnqualifiedType().getCanonicalType()` .  

I'll poke around for way to get the name of `BuiltinType::Float128` from somewhere not unreasonable.

Reply:
`S.Context.Float128Ty` for the win.